vllm/model_executor/layers/logits_processor.py

```diff
@@ -76,7 +80,15 @@ def _get_logits(self, hidden_states: torch.Tensor,
         logits = lm_head.linear_method.apply(lm_head,
                                              hidden_states,
                                              bias=embedding_bias)
-        logits = tensor_model_parallel_gather(logits)
+        if self.use_gather:
+            logits = tensor_model_parallel_gather(logits)
+        else:
+            # Gather is not supported for some devices such as TPUs.
+            # Use all-gather instead.
+            # NOTE(woosuk): Here, the outputs of every device should not be None
+            # because XLA requires strict SPMD among all devices. Every device
+            # should execute the same operations after gathering the logits.
+            logits = tensor_model_parallel_all_gather(logits)
```

Comment:
ideally, this dispatching logic should not exist in logits processors, but handled by TPU communicators.

Reply:
Again, this is hard because "gather" means outputting None in non-root devices.

Reply:
sorry i don't understand. per my understanding, you can use `allgather`, and then discard the output in non-root devices.

Reply:
As written in the comment, XLA requires strict SPMD. The graph executed by the devices should all be the exactly same. If the logits is None in non-root devices, they will have a graph that does not execute the following lines of code such as `logits = logits[:, :self.org_vocab_size]` and sampler. This will cause the difference between the graph for the root device, which XLA does not allow.